devito/data/utils.py

```diff
@@ -0,0 +1,259 @@
+        else:
+            stop = -(idx.stop % (modulo + 1))
+        return slice(start, stop, idx.step)
+    elif isinstance(idx, (tuple, list)):
+        return [i % modulo for i in idx]
+    elif isinstance(idx, np.ndarray):
+        return idx
+    else:
+        raise ValueError("Cannot apply modulo to index of type `%s`" % type(idx))
+
+
+def index_dist_to_repl(idx, decomposition):
+    """Convert a distributed array index into a replicated array index."""
+    if decomposition is None:
+        return PROJECTED if is_integer(idx) else slice(None)
+
+    # Derive shift value
+    if isinstance(idx, slice):
+        if idx.step is None or idx.step >= 0:
+            value = idx.start
+        else:
+            value = idx.stop
+    else:
+        value = idx
+    if value is None:
+        value = 0
+    elif not is_integer(value):
+        raise ValueError("Cannot derive shift value from type `%s`" % type(value))
+
+    # Convert into absolute local index
+    idx = decomposition.index_glb_to_loc(idx, rel=False)
+
+    if is_integer(idx):
+        return PROJECTED
+    elif idx is None:
+        return NONLOCAL
+    elif isinstance(idx, (tuple, list)):
+        return [i - value for i in idx]
+    elif isinstance(idx, np.ndarray):
+        return idx - value
+    elif isinstance(idx, slice):
+        if idx.step is not None and idx.step < 0:
+            if idx.stop is None:
+                return slice(idx.start - value, None, idx.step)
+        return slice(idx.start - value, idx.stop - value, idx.step)
+    else:
+        raise ValueError("Cannot apply shift to type `%s`" % type(idx))
+
+
+def convert_index(idx, decomposition, mode='glb_to_loc'):
+    """Convert a global index into a local index or vise versa according to mode."""
+    if is_integer(idx) or isinstance(idx, slice):
+        return decomposition(idx, mode=mode)
+    elif isinstance(idx, (tuple, list)):
+        return [decomposition(i, mode=mode) for i in idx]
+    elif isinstance(idx, np.ndarray):
+        return np.vectorize(lambda i: decomposition(i, mode=mode))(idx)
+    else:
+        raise ValueError("Cannot convert index of type `%s` " % type(idx))
+
+
+def index_handle_oob(idx):
+    # distributed.index_glb_to_loc returns None when the index is globally
+    # legal, but out-of-bounds for the calling MPI rank
+    if idx is None:
+        return NONLOCAL
+    elif isinstance(idx, (tuple, list)):
+        return [i for i in idx if i is not None]
+    elif isinstance(idx, np.ndarray):
+        if idx.dtype == np.bool_:
+            # A boolean mask, nothing to do
+            return idx
+        elif idx.ndim == 1:
+            return np.delete(idx, np.where(idx == None))  # noqa
+        else:
+            raise ValueError("Cannot identify OOB accesses when using "
+                             "multidimensional index arrays")
+    else:
+        return idx
+
+
+def local_data_idx(loc_idx):
+    """Retrieve the unflipped idx corresponding to loc_idx."""
```

Comment:
can you elaborate on what you mean by 'unflipped' here? 

either here or where it's used for the first time as jargon

Reply:
do we have a test specifically for this function? or is it just tested indirectly?

Reply:
Yes, this has a specific test. Will elaborate on 'unflipped'.